src/common/types/vector.cpp

```diff
@@ -298,6 +298,13 @@ void Vector::SetValue(idx_t index, const Value &val) {
 		auto &child = DictionaryVector::Child(*this);
 		return child.SetValue(sel_vector.get_index(index), val);
 	}
+	if (val.type().id() == LogicalTypeId::DECIMAL && GetType().id() == LogicalTypeId::DECIMAL) {
+		//! Ensure that both are of the exact same DECIMAL type
+		if (val.type() != GetType()) {
+			SetValue(index, val.CastAs(GetType()));
+			return;
+		}
+	}
 	if (val.type().InternalType() != GetType().InternalType()) {
```

Comment:
I figured that would be more expensive to do for every SetValue call, so I only added it for the case I knew was problematic, but I'll change it

Reply:
Hmm because of this change a test is failing
```c++
	result.emplace_back(LogicalType::BLOB, "blob", Value::BLOB("thisisalongblob\\x00withnullbytes"),
	                    Value("\\x00\\x00\\x00a"));
```
The last value was previously not being cast, because VARCHAR and BLOB share the same internal type
But now that we are casting them, this no longer remains escaped
Previous result:
```py
[(b'thisisalongblob\x00withnullbytes',), (b'\\x00\\x00\\x00a',), (None,)]
```
New result
```py
[(b'thisisalongblob\x00withnullbytes',), (b'\x00\x00\x00a',), (None,)]
```

Reply:
Looks to me like an oversight actually, that should be `Value::BLOB` there as well I believe

Reply:
I'm pretty sure this is a bug
`\\` should get escaped down to `\`, resulting in the contents of the string being:
`['\', '\0', '\', '\0', '\', '\0', 'a']`
If we convert that back to BLOB every character should be interpreted separately, resulting in the same contents (so essentially no cast)

The issue probably stems from using the same cast for string values from the queries as the VARCHAR values inside the database
```sql
duckdb> select '\x00\x00\x00a'::BLOB;
┌───────────────────────┐
│ '\x00\x00\x00a'::BLOB │
╞═══════════════════════╡
│ 00000061              │
└───────────────────────┘
```

Reply:
This is not a bug but intended behavior, for blobs `\x00` is an escape code that means `the byte 0`. The resulting blob in the example (`\x00\x00\x00a`) should be 4 bytes, `[0, 0, 0, 97]`. Previously the string `\x00\x00\x00a` was stored as a string, but silently converted to a blob because of the missing type cast. This caused the contents of the blob to be the ascii values of that string instead of handling the escape code, i.e. the blob used to have 13 characters with as content `[92, 120, 48, 48, 92, 120, 48, 48, 92, 120, 48, 48, 97]`. The new result seems correct to me as what was intended was likely to use the hex escape codes to create a blob with null bytes.

Reply:
Ah, I see the confusion now, the escaping double back-slash is to escape the C++ string. The C++ string `"\\x00" gets converted into the actual string "\x00". In order to escape the blob a double escaped backslash would be necessary (i.e. "\\\\x00").

Reply:
I did find another issue possibly caused by this
Observing a crash in `test_alias_table`
Appears to be an infinite recursion, looks like it's caused by this change
```
* thread #1, queue = 'com.apple.main-thread', stop reason = breakpoint 1.1
    frame #0: 0x000000010c10633c libduckdb.dylib`duckdb::Vector::SetValue(this=0x0000600003700960, index=0, val=0x000000016fdf7df0) at vector.cpp:302:12
   299                  return child.SetValue(sel_vector.get_index(index), val);
   300          }
   301          if (val.type() != GetType()) {
-> 302                  SetValue(index, val.CastAs(GetType()));
   303                  return;
   304          }
   305 
(lldb) p val.type()
(const duckdb::LogicalType) $4 = {
  id_ = VARCHAR
  physical_type_ = VARCHAR
  type_info_ = nullptr {
    __ptr_ = nullptr
  }
}
(lldb) p GetType()
(const duckdb::LogicalType) $5 = {
  id_ = VARCHAR
  physical_type_ = VARCHAR
  type_info_ = std::__1::shared_ptr<duckdb::ExtraTypeInfo>::element_type @ 0x000060000211c798 strong=19 weak=1 {
    __ptr_ = 0x000060000211c798
  }
}
(lldb) n
Process 84474 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = breakpoint 1.1
    frame #0: 0x000000010c10633c libduckdb.dylib`duckdb::Vector::SetValue(this=0x0000600003700960, index=0, val=0x000000016fdf7b50) at vector.cpp:302:12
   299                  return child.SetValue(sel_vector.get_index(index), val);
   300          }
   301          if (val.type() != GetType()) {
-> 302                  SetValue(index, val.CastAs(GetType()));
   303                  return;
   304          }
   305 
(lldb) p GetType()
(const duckdb::LogicalType) $6 = {
  id_ = VARCHAR
  physical_type_ = VARCHAR
  type_info_ = std::__1::shared_ptr<duckdb::ExtraTypeInfo>::element_type @ 0x000060000211c798 strong=19 weak=1 {
    __ptr_ = 0x000060000211c798
  }
}
(lldb) p val.type()
(const duckdb::LogicalType) $7 = {
  id_ = VARCHAR
  physical_type_ = VARCHAR
  type_info_ = nullptr {
    __ptr_ = nullptr
  }
}
```

Reply:
Ah, this is caused by a recent PR. The problem seems to be that the varchar type has an alias, but `Vector::GetValue` does not correctly set the alias of the value. Try rewriting `Vector::GetValue` to the following:

```cpp
Value Vector::GetValue(idx_t index) const {
    auto result = GetValue(*this, index);
    result.type().SetAlias(GetType().GetAlias());
    return result;
}
```